### PR TITLE
Added taskgroup decorator

### LINKED ISSUE #11870: Add @taskgroup decorator
**Description**

Similarly to `@dag` and `@task` decorators, we can add a decorator to easily generate `TaskGroup` instances. 

**Use case / motivation**

This would be used for reusable pieces of DAGs. An example may be to hyperparameter tune an ML pipeline. You may want to generate several parallel subpipelines with different settings/parameters to be executed in parallel. At the end pull the result and decide the best parametrization.  

```py
@task
collect_dataset(...)
  pass

@task
def train_model(...)
   pass

@task_group
def movielens_model_pipeline(learning_rate: int, feature_list: list, dataset: XComArg):  
  dataset = filter_dataset(dataset, feature_list)
  train_model(dataset, learning_rate)

@dag
def movielens_hpt(dataset_path: str, feature_list:list=['year', 'director']):
  dataset = load_dataset(dataset_path)
  for i in range(0.1, 0.9, step=0.1):
    movielens_model_pipeline(i, feature_list, dataset)
  
  decide_best_model(...)
```

**Related Issues**

https://github.com/apache/airflow/issues?q=is%3Aopen+is%3Aissue+label%3AAIP-31

### PR DESCRIPTION
**Pull request by VBhojawala**
Added taskgroup decorator which can be used to create taskgroup from python callable. Inside python callable tasks can be grouped by calling tasks (python callable ) created with task decorator.

- taskgroup decorator takes both optional and keyword argument which are passed to Constructor of TaskGroup class.

- TaskGroup can be created with one of following syntax

``` python
@taskgroup
@taskgroup()
@taskgroup(group_id='group_name')
```

- TaskGroup class  constructor takes one mandatory argument group_id, if not given in decorator it sets group_id to python callable name.

Following is a simple example demonstrating use of taskgroup decorator grouping multiple tasks.

``` python
@task
def task_1(value):
  return f'[ Task1 {value} ]'


@task
def task_2(value):
  print(f'[ Task2 {value} ]')


@taskgroup
def section_1(value):
  return task_2(task_1(value))
```

taskgroup decorator utilizes existing TaskGroup context manager currently used for creating TaskGroup, which means we can create nested taskgroup by created nested callable. Following is an example demonstrating use of nested taskgroup.

``` python
@task
def task_start():
    return '[Task_start]'

@task
def task_end():
    print(f'[ Task_End ]')

@task
def task_1(value):
    return f'[ Task1 {value} ]'

@task
def task_2(value):
    print(f'[ Task2 {value} ]')

@task
def task_3(value):
    return f'[ Task3 {value} ]'

@task
def task_4(value):
    print(f'[ Task4 {value} ]')

@taskgroup
def section_1(value):

    @taskgroup
    def section_2(value2):
        return task_4(task_3(value2))

    op1 = task_2(task_1(value))
    return section_2(op1) 
```

Dedicated test cases for taskgroup decorator is created in file 
/tests/utils/test_task_group_decorator.py

Recent changes 

- Added logic to append suffix to duplicate group_id.

closes: #11870

### CODE REVIEW

**Comment by VBhojawala** [comment]
Hi @casassg ,

I have raised new pull request with suggested changes. Kindly review it.

File: airflow/utils/task_group.py
Comment (ashb) [inline airflow/utils/task_group.py]: Typo

Comment (VBhojawala) [reply airflow/utils/task_group.py]: Fixed the Typo.

**Review by casassg — APPROVED** [review]
this looks great! Thanks for taking your time to contribute this!

a small nit: Could you add an extra test that calls a taskgroup decorated function 2 times.


```py
@taskgroup
def training(paths):
  validate(train(paths))

@dag
def pipeline():
  training('some/path')
  training('some/other/path')
```

**Review by casassg — APPROVED** [review]
ty! This looks good to me. Will need a committer for a final review

**Comment by kaxil** [comment]
Can you rebase the PR on latest master please

**Comment by VBhojawala** [comment]
Hi @kaxil, I have rebased the  PR.

File: airflow/utils/task_group.py
Comment (kaxil) [inline airflow/utils/task_group.py]: We don't want a print statement here I think -- (maybe a remnant of debugging)

Comment (VBhojawala) [reply airflow/utils/task_group.py]: ya,

File: airflow/utils/task_group.py
Comment (turbaszek) [inline airflow/utils/task_group.py:127]: This feature allow users to re use the same `TaskGroup` in the DAG, right?

File: docs/apache-airflow/concepts.rst
Comment (turbaszek) [inline docs/apache-airflow/concepts.rst]:
```suggestion
TaskGroup can be created using ``@taskgroup`` decorator, it takes one mandatory argument ``group_id`` which is same as constructor of TaskGroup class, if not given it copies function name as ``group_id``. It works exactly same as creating TaskGroup using context manager ``with TaskGroup('groupid') as section:``.
```

File: docs/apache-airflow/concepts.rst
Comment (turbaszek) [inline docs/apache-airflow/concepts.rst]: Should we use the `exampleinclude` to use code from example DAg instead of hardcoding it in docs?

```
.. exampleinclude:: /../../airflow/providers/apache/beam/example_dags/example_beam.py
    :language: python
    :dedent: 4
    :start-after: [START howto_operator_start_python_dataflow_runner_pipeline_gcs_file]
    :end-before: [END howto_operator_start_python_dataflow_runner_pipeline_gcs_file]
```

**Review by turbaszek — APPROVED** [review]

File: airflow/utils/task_group.py
Comment (VBhojawala) [reply airflow/utils/task_group.py:127]: Correct

**Review by turbaszek — APPROVED** [review]

File: airflow/utils/task_group.py
Comment (ashb) [inline airflow/utils/task_group.py]:
```suggestion
from inspect import signature
```

Please import directly form the module, not from somewhere else where we us it.

File: tests/utils/test_task_group.py
Comment (ashb) [inline tests/utils/test_task_group.py:837]: I am not sure this is the right behavour!

This is _separate_ task groups with the same id. That feels like a logic bug on the DAG author part and should throw an error.

What do others think of this?

**Review by ashb — CHANGES_REQUESTED** [review]

File: tests/utils/test_task_group.py
Comment (casassg) [reply tests/utils/test_task_group.py:837]: I agree. But we can't retrieve the function for previously called decorated functions, which means that we can't know wether the function name is the same or not (aka wether its the same function being called 2 times or 2 functions w the same group_id). So while I agree this is an unintended side effect, I would leave it as is as it allows us to instatiate the same function several times

Comment (ashb) [reply tests/utils/test_task_group.py:837]: We can't check what task group decorator created the previous function, but we could check if the _current_ TG was created by the current decorated function, and if not then it would fail.

We could keep a set of TG objects in the wrapper fn, and if the TG already exists on the DAG, and the TG is not in that set, then it was created by some other decorator -> raise an error.

Comment (casassg) [reply tests/utils/test_task_group.py:837]: Makes sense. I don't think it's a big deal either way. If users do this, even though it's weird and can be confusing, they can do it. 

Mostly want to make sure we are valuing the right trade offs between complexity added and value of having this extra check.

Not strongly opinionated though

Comment (ashb) [reply tests/utils/test_task_group.py:837]: My thought was that this is more likely a bug than desired behaviour -- for instance someone copy-and-pasted a task group but forgot to update the name

File: airflow/example_dags/example_task_group_decorator.py
Comment (dimberman) [inline airflow/example_dags/example_task_group_decorator.py:59]: might be nicer to call it "@task_group" instead of "@taskgroup" since there's no capitalization

File: airflow/example_dags/example_task_group_decorator.py
Comment (dimberman) [inline airflow/example_dags/example_task_group_decorator.py:62]: Can you unwind this? I wound't want to encourage people to nest like this.

**Review by dimberman — CHANGES_REQUESTED** [review]

File: tests/utils/test_task_group.py
Comment (dimberman) [reply tests/utils/test_task_group.py:837]: I agree with @casassg here. It would be very hard to enforce this. I'm especially imagining situations where users import taskgroups from libraries and therefore do not have control over the task_group_id. That said, it would make sense for us to allow users to override the task_group_id via a kwarg in case they import two functions with the same id name.

**Comment by dimberman** [comment]
Hi @VBhojawala are you interested in continuing this PR? I'm glad to either finish it for you or help you finish :)

**Comment by VBhojawala** [comment]
Hi @dimberman ,

Kindly review below PR and help me to finish it : 
https://github.com/apache/airflow/pull/13405

related: #8970

**Comment by ashb** [comment]
> Hi @dimberman ,
> 
> Kindly review below PR and help me to finish it :
> #13405
> 
> related: #8970

@VBhojawala commenting on another unrelated PR asking for a review is considered poor etiquette - please don't do this.

**Comment by ashb** [comment]
Closing in favour of #15034 which continues this work.

**Pull request closed without merging**
